r/tests/testthat/test-dplyr-funcs-datetime.R

```diff
@@ -711,3 +711,40 @@ test_that("am/pm mirror lubridate", {
   )
 
 })
+test_that("dst extracts daylight savings time correctly", {
+  test_df <- tibble(
+    dates = as.POSIXct(c("2021-02-20", "2021-07-31", "2021-10-31", "2021-01-31"), tz = "Europe/London")
+  )
+  # https://issues.apache.org/jira/browse/ARROW-13168
+  skip_on_os("windows")
+
+  compare_dplyr_binding(
+    .input %>%
+      mutate(dst = dst(dates)) %>%
+      collect(),
+    test_df
+  )
+})
+
+test_that("dst errors with unsupported input", {
+  expect_error(
+    call_function("is_dst", Scalar$create("this is a string, not a timestamp")),
+    "NotImplemented: Function 'is_dst' has no kernel matching input types (scalar[string])",
+    fixed = TRUE
+  )
+  expect_error(
+    call_function("is_dst", Scalar$create(1L)),
+    "NotImplemented: Function 'is_dst' has no kernel matching input types (scalar[int32])",
+    fixed = TRUE
+  )
+  expect_error(
+    call_function("is_dst", Scalar$create(2.2)),
+    "NotImplemented: Function 'is_dst' has no kernel matching input types (scalar[double])",
+    fixed = TRUE
+  )
+  expect_error(
+    call_function("is_dst", Scalar$create(TRUE)),
+    "NotImplemented: Function 'is_dst' has no kernel matching input types (scalar[bool])",
+    fixed = TRUE
+  )
```

Comment:
I think, in this case, the error messages are fine. I think they are good enough. However, and this is in relation with your follow-up questions, I don't think they are very consistent. Take for example, calling the `"date"` binding on a double:
```r
> call_binding("date", Scalar$create(32.2))
 Error: NotImplemented: Unsupported cast from double to date32 using function cast_date32
```
Some bindings are implemented at a higher and some at a lower level. Do we want the error messaging to be consistent (or, better said, as consistent as it could be)? If the answer is "Yes", then we might need to move that to R instead of relying on the C++ code. Other benefits (until cpp11 supports classed errors): handling errors might be a bit easier.

Reply:
We might need to write some tricky bindings though (`tryCatch()` comes to mind).

Reply:
I think you make a great point here @dragosmg about consistent error messaging.  I just wonder if the term "kernel" might be unfamiliar to our R users who are less familiar with the C++ underpinnings of the package as well.

Reply:
I'd say it's a pretty good chance _kernel_ is an unfamiliar term to the regular R user.

Reply:
In which case I'd say that gives us good reason to write our own error messages here.

Reply:
While I totally agree `"kernel"` is likely to be unfamiliar — there is other content in this message that IMO conveys what's going on. Do we think that `"kernel"` will confuse people enough that they wouldn't understand the rest of the message or stump them such that they don't look at the rest of the message? I guess my approach for something like this would be trying to read the rest of the message and it's (not a grammatical) sentence, but does describe it in  basic ways (albeit our type info has some extra stuff about array/scalar).

I'm legit on the fence myself about this, I don't know the right answer here! If we do want to smooth these out and remove the word `kernel` I would suggest we do that as a separate Jira so we can (1) get a few more eyes on the decision and people are aware of it going forward. (2) we likely have _many_ circumstances where messages like these show up in the package (and many of those aren't tested since we don't test things directly from C++). So we will want to do something generally to approach this (we don't necessarily want to add type-checking in front of every C++ call in the package and writing out a new message each time we do that)

Reply:
I agree with you @jonkeane both on term `"kernel"` and on a Jira for a more general improvement of error messages.